src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs

```diff
@@ -0,0 +1,260 @@
+                    if (delete.Validate(context, items, index) == false)
+                    {
+                        return false;
+                    }
+                    toExecute.Add(delete);
+                }
+            }
+
+            foreach (var command in toExecute)
+            {
+                command.Execute(context, items, index);
+            }
+
+            return true;
+        }
+
+        public unsafe void SaveCommandsBatch(TransactionOperationContext context, long index)
+        {
+            var items = context.Transaction.InnerTransaction.OpenTable(ClusterStateMachine.TransactionCommandsSchema, ClusterStateMachine.TransactionCommands);
+            var commands = SerializedDatabaseCommands.Clone(context);
+            using (Slice.From(context.Allocator, Database + "/", out Slice databaseSlice))
+            using (items.Allocate(out TableValueBuilder tvb))
+            {
+                tvb.Add(databaseSlice.Content.Ptr, databaseSlice.Size);
+                tvb.Add(Bits.SwapBytes(index));
+                tvb.Add(commands.BasePointer, commands.Size);
+                items.Insert(tvb);
+            }
+        }
+
+        public enum TransactionCommandsColumn
+        {
+            Database,
+            RaftIndex,
+            Commands,
+        }
+
+        public static unsafe long ReadFirstIndex(TransactionOperationContext serverContext, string database)
+        {
+            var items = serverContext.Transaction.InnerTransaction.OpenTable(ClusterStateMachine.TransactionCommandsSchema, ClusterStateMachine.TransactionCommands);
+
+            using (Slice.From(serverContext.Allocator, database + "/", out var prefixSlice))
+            {
+                var reader = items.SeekOneForwardFrom(ClusterStateMachine.TransactionCommandsSchema.Indexes[ClusterStateMachine.CommandByDatabaseAndIndex], prefixSlice);
+                if (reader == null)
+                    return 0;
+                var value = *(long*)reader.Reader.Read((int)TransactionCommandsColumn.RaftIndex, out var _);
+                return Bits.SwapBytes(value);
+            }
+        }
+
+        public static (BlittableJsonReaderArray Commands, ClusterTransactionOptions Options) ReadCommandsBatch(TransactionOperationContext context, string database, long index)
+        {
+            var items = context.Transaction.InnerTransaction.OpenTable(ClusterStateMachine.TransactionCommandsSchema, ClusterStateMachine.TransactionCommands);
+
+            var databaseBytes = Encoding.UTF8.GetBytes(database + "/");
+            var indexBytes = BitConverter.GetBytes(Bits.SwapBytes(index));
+            
+            var prefix = new byte[databaseBytes.Length + sizeof(long)];
+            databaseBytes.CopyTo(prefix, 0);
+            indexBytes.CopyTo(prefix, databaseBytes.Length);
+
+            using (Slice.From(context.Allocator, prefix, out Slice prefixSlice))
```

Comment:
This uses a lot of allocations that aren't required. Can do that directly on a buffer.